Added JSON output to cli commands

petarjuki7 opened the pull request:
You can now add the `--test` or `-t` flag to `meroctl` commands and receive outputs in JSON format.

miraclx reviewed (commented):
why not `--json`, `--format json` or something of the sort?

petarjuki7 commented:
> why not `--json`, `--format json` or something of the sort?

Sure, makes sense

chefsale commented:
Status on this PR, is it still valid?

chefsale commented:
To rephrase this question then: Is this still in progress and whats the ETA? :)

petarjuki7 commented:
This PR increased in scope a bit so I'm going to explain the changes.

1.  Added `context update` command
2.  The functions in the `context create` file now use the commands from the `app` part of the cli, it made sense to me to reuse the code because the way it was before seemed like duplicating a lot of the code
3. Most important change, removed `eyre` from most of the crates, now we have a custom error type. Influenced by the way it was done in [ws.rs](https://github.com/calimero-network/core/blob/master/crates/server/src/ws.rs#L351-L377) and [jsonrpc.rs](https://github.com/calimero-network/core/blob/master/crates/server/src/jsonrpc.rs#L114-L140), we push the responses to the top level and then print out just one JSON

Open to feedback on `error` naming and different kinds of errors, the way it is right now is mostly influenced by the `ws.rs` file, so everything may not fit perfectly into the cli

cc: @miraclx  @fbozic

miraclx commented on crates/meroctl/src/cli/app/install.rs:
```diff
@@ -30,22 +30,34 @@ pub struct InstallCommand {
             is_dev_installation = true;
-            serde_json::to_value(install_dev_request)?
+            serde_json::to_value(install_dev_request)
+                .map_err(|e| CliError::InternalError(format!("{}", e.to_string())))?
```
```suggestion
                .map_err(|e| CliError::InternalError(e.to_string()))?
```

et al.

miraclx commented on crates/meroctl/src/cli/app/install.rs:
```diff
@@ -30,22 +30,34 @@ pub struct InstallCommand {
+            let install_dev_request = InstallDevApplicationRequest::new(
+                app_path
+                    .canonicalize_utf8()
+                    .map_err(|_| CliError::InternalError(format!("Canonicalize path failed")))?,
```
```suggestion
                    .map_err(|_| CliError::InternalError("Canonicalize path failed".to_owned()))?,
```

et al.

miraclx commented on crates/meroctl/src/cli/context/create.rs:
```diff
@@ -76,56 +80,55 @@ impl CreateCommand {
-        bail!("Application is not installed on node.")
+    args: &RootArgs,
+) -> Result<CreateContextResponse, CliError> {
+    if app_installed(&application_id, &args).await.is_err() {
```
quick check: is it supposed to return an error if the app doesn't exist?

miraclx commented on crates/meroctl/src/cli/context/watch.rs:
```diff
@@ -15,39 +15,49 @@ pub struct WatchCommand {
             match message {
                 Ok(msg) => {
                     if let tokio_tungstenite::tungstenite::Message::Text(text) = msg {
-                        println!("{}", text);
+                        println!(
+                            "{:#?}",
```
bummer, these won't fit the model

miraclx reviewed (commented):
`context watch` not being able to conform to the rest of the app suggests perhaps a better model would be similar to how rust does it

```
$ cargo build -p meroctl --message-format json
{ json message 1 }
{ json message 2 }
{ json message 3 }
```

anything that produces a console report would emit a report instead of returning one

depending on the presence of the `--format` flag, we can either catch that report and print a pretty console output or print the JSON serialized form of the individual messages

https://docs.rs/cargo_metadata/latest/cargo_metadata/index.html

MatejVukosav commented:
Please add this to user docs when ready
